quantum/painter/qp_draw_image.c

```diff
@@ -260,15 +262,24 @@ static bool qp_drawimage_recolor_impl(painter_device_t device, uint16_t x, uint1
         return false;
     }
 
-    // Set up the output state
-    struct qp_internal_pixel_output_state output_state = {.device = device, .pixel_write_pos = 0, .max_pixels = qp_internal_num_pixels_in_buffer(device)};
+    bool ret = false;
+    if (frame_info->bpp <= 8) {
+        // Set up the output state
+        struct qp_internal_pixel_output_state output_state = {.device = device, .pixel_write_pos = 0, .max_pixels = qp_internal_num_pixels_in_buffer(device)};
 
-    // Decode the pixel data and stream to the display
-    bool ret = qp_internal_decode_palette(device, pixel_count, frame_info->bpp, input_callback, &input_state, qp_internal_global_pixel_lookup_table, qp_internal_pixel_appender, &output_state);
+        // Decode the pixel data and stream to the display
+        ret = qp_internal_decode_palette(device, pixel_count, frame_info->bpp, input_callback, &input_state, qp_internal_global_pixel_lookup_table, qp_internal_pixel_appender, &output_state);
+        // Any leftovers need transmission as well.
+        if (ret && output_state.pixel_write_pos > 0) {
+            ret &= driver->driver_vtable->pixdata(device, qp_internal_global_pixdata_buffer, output_state.pixel_write_pos);
+        }
+    } else {
+        // Set up the output state
+        struct qp_internal_byte_output_state output_state = {.device = device, .byte_write_pos = 0, .max_bytes = QUANTUM_PAINTER_PIXDATA_BUFFER_SIZE};
```

Comment:
```suggestion
        struct qp_internal_byte_output_state output_state = {.device = device, .byte_write_pos = 0, .max_bytes = qp_internal_num_pixels_in_buffer(device) * driver->native_bits_per_pixel / 8};
```
Using just `QUANTUM_PAINTER_PIXDATA_BUFFER_SIZE` may result in partial pixels (e.g., in the RGB888 case, if the buffer size is some usual power of 2), and the `pixdata` method which actually sends the data to the device does not support partial pixels.

Reply:
I first checked the datasheet for only display in QP that works with RGB888 and found that it actually allows partial pixel data to be sent...but then I re-read your comment and realized the "full pixel" constraint is with `pixdata` itself, so you are correct; I've incorporated the recommended change.